semgrep-core/src/parsing/Parse_rule.ml

```diff
@@ -414,144 +431,110 @@ let parse_languages ~id langs =
+           ( Container
+               ( Dict,
+                 (_, [ Tuple (_, [ L (String ("rules", _)); rules ], _) ], _) ),
+             _ );
+       _;
+     };
+    ] ->
+        rules
+    | _ -> error "missing rules entry as top-level key"
+  in
+  let rules =
+    match rules_block with
+    | Container (Array, (_, rules, _)) -> rules
+    | _ -> error "expected a list of rules following `rules:`"
+  in
+  rules
+  |> List.mapi (fun i rule ->
+         let rule_dict = yaml_to_dict "rules" rule in
+         let take f key = take rule_dict f key in
+         let take_opt f key = take_opt rule_dict f key in
+         let id, languages =
+           (take parse_string "id", take parse_string_list "languages")
+         in
+         let languages = parse_languages ~id languages in
+         let env = { id; languages; path = [ string_of_int i; "rules" ] } in
+         let ( message,
+               severity,
+               mode_opt,
+               metadata_opt,
+               fix_opt,
+               fix_regex_opt,
+               paths_opt,
+               equivs_opt,
+               options_opt ) =
+           ( take parse_string "message",
+             take parse_string "severity",
+             take_opt parse_string "mode",
+             take_opt generic_to_json "metadata",
+             take_opt parse_string "fix",
+             take_opt (parse_fix_regex env) "fix-regex",
+             take_opt parse_paths "paths",
+             take_opt parse_equivalences "equivalences",
+             take_opt parse_options "options" )
+         in
+         let mode = parse_mode env mode_opt rule_dict in
+         {
+           R.id;
+           message;
+           languages;
+           file;
+           severity = H.parse_severity ~id severity;
+           mode;
+           (* optional fields *)
+           metadata = metadata_opt;
+           fix = fix_opt;
+           fix_regexp = fix_regex_opt;
+           paths = paths_opt;
+           equivalences = equivs_opt;
+           options = options_opt;
+         })
 
 let parse file =
-  let json =
+  let ast =
     match FT.file_type_of_file file with
-    | FT.Config FT.Yaml -> (
-        let str = Common.read_file file in
-        let yaml_res = Yaml.of_string str in
-        match yaml_res with
-        | Result.Ok v -> yaml_to_json v
-        | Result.Error (`Msg s) -> raise (E.UnparsableYamlException s))
-    | FT.Config FT.Json -> J.load_json file
+    | FT.Config FT.Yaml -> Yaml_to_generic.program file
+    | FT.Config FT.Json ->
+        Json_to_generic.program (Parse_json.parse_program file)
     | FT.Config FT.Jsonnet ->
         Common2.with_tmp_file ~str:"parse_rule" ~ext:"json" (fun tmpfile ->
             let cmd = spf "jsonnet %s -o %s" file tmpfile in
             let n = Sys.command cmd in
             if n <> 0 then failwith (spf "error executing %s" cmd);
-            J.load_json tmpfile)
+            Json_to_generic.program (Parse_json.parse_program tmpfile))
```

Comment:
Do we really need to do the conversion to Generic? What is the advantage?

Reply:
I tried to explain in the "history:" comment at the top of this file.  Basically we need that to get precise location information.